File: src/ormai/adapters/sqlalchemy/introspection.py
"""
SQLAlchemy schema introspection.

Extracts metadata from SQLAlchemy models to build the schema representation.
"""

from typing import Any

from sqlalchemy import inspect
from sqlalchemy.orm import RelationshipProperty
from sqlalchemy.sql.sqltypes import (
    Boolean,
    Date,
    DateTime,
    Float,
    Integer,
    LargeBinary,
    String,
    Text,
    Time,
)

from ormai.core.types import (
    FieldMetadata,
    FieldType,
    ModelMetadata,
    RelationMetadata,
    SchemaMetadata,
)


class SQLAlchemyIntrospector:
    """
    Introspects SQLAlchemy models to extract schema metadata.
    """

    def __init__(self, models: list[type]) -> None:
        """
        Initialize with a list of SQLAlchemy model classes.

        Args:
            models: List of SQLAlchemy declarative model classes
        """
        self.models = models
        self._model_map: dict[str, type] = {
            self._get_model_name(m): m for m in models
        }

    def introspect(self) -> SchemaMetadata:
        """
        Introspect all registered models and return schema metadata.
        """
        models_meta: dict[str, ModelMetadata] = {}

        for model in self.models:
            model_name = self._get_model_name(model)
            models_meta[model_name] = self._introspect_model(model)

        return SchemaMetadata(models=models_meta)

    def _get_model_name(self, model: type) -> str:
        """Get the name to use for a model."""
        return model.__name__

    def _introspect_model(self, model: type) -> ModelMetadata:
        """Introspect a single model."""
        mapper = inspect(model)

        # Get table name
        table_name = mapper.mapped_table.name

        # Get fields
        fields: dict[str, FieldMetadata] = {}
        primary_keys: list[str] = []

        for column in mapper.columns:
            field_meta = self._introspect_column(column)
            fields[column.key] = field_meta
            if column.primary_key:
                primary_keys.append(column.key)

        # Get relations
        relations: dict[str, RelationMetadata] = {}
        for rel in mapper.relationships:
            rel_meta = self._introspect_relationship(rel)
            relations[rel.key] = rel_meta

        return ModelMetadata(
            name=self._get_model_name(model),
            table_name=table_name,
            fields=fields,
            relations=relations,
            primary_keys=primary_keys,
            description=model.__doc__,
        )

    def _introspect_column(self, column: Any) -> FieldMetadata:
        """Introspect a single column."""
        return FieldMetadata(
            name=column.key,
            field_type=self._get_field_type(column.type),
            nullable=column.nullable or False,
            primary_key=column.primary_key,
            default=self._get_default(column),
            description=column.doc,
        )

    def _introspect_relationship(self, rel: RelationshipProperty) -> RelationMetadata:
        """Introspect a relationship."""
        # Determine relationship type
        if rel.uselist:
            if rel.secondary is not None:
                rel_type = "many_to_many"
            else:
                rel_type = "one_to_many"
        else:
            rel_type = "many_to_one"

        # Get foreign key if available
        foreign_key = None
        if rel.local_columns:
            fk_cols = list(rel.local_columns)
            if fk_cols:
                foreign_key = fk_cols[0].key

        return RelationMetadata(
            name=rel.key,
            target_model=rel.mapper.class_.__name__,
            relation_type=rel_type,
            foreign_key=foreign_key,
            back_populates=rel.back_populates,
        )

    def _get_field_type(self, sa_type: Any) -> FieldType:
        """Map SQLAlchemy type to OrmAI FieldType."""
        type_mapping = {
            String: FieldType.STRING,
            Text: FieldType.STRING,
            Integer: FieldType.INTEGER,
            Float: FieldType.FLOAT,
            Boolean: FieldType.BOOLEAN,
            DateTime: FieldType.DATETIME,
            Date: FieldType.DATE,
            Time: FieldType.TIME,
            LargeBinary: FieldType.BINARY,
        }

        for sa_class, field_type in type_mapping.items():
            if isinstance(sa_type, sa_class):
                return field_type

        # Check for UUID
        type_name = type(sa_type).__name__.lower()
        if "uuid" in type_name:
            return FieldType.UUID
        if "json" in type_name:
            return FieldType.JSON

        return FieldType.UNKNOWN

    def _get_default(self, column: Any) -> Any:
        """Extract default value from column if available."""
        if column.default is None:
            return None
        if column.default.is_scalar:
            return column.default.arg
        # For callable defaults, we can't represent them
        return None

    def get_model_class(self, name: str) -> type | None:
        """Get the model class by name."""
        return self._model_map.get(name)

File: ormai-ts/src/adapters/drizzle/introspection.ts
/**
 * Drizzle ORM schema introspection.
 */

import type {
  FieldMetadata,
  FieldType,
  ModelMetadata,
  RelationMetadata,
  SchemaMetadata,
} from '../../core/types.js';

/**
 * Drizzle table type.
 */
export interface DrizzleTable {
  [key: string]: unknown;
  _: {
    name: string;
    columns: Record<string, DrizzleColumn>;
  };
}

/**
 * Drizzle column type.
 */
export interface DrizzleColumn {
  name: string;
  dataType: string;
  notNull: boolean;
  primary?: boolean;
  default?: unknown;
  columnType?: string;
}

/**
 * Drizzle relation definition.
 */
export interface DrizzleRelation {
  referencedTableName: string;
  fieldName: string;
  relationType: 'one' | 'many';
}

/**
 * Drizzle schema type (record of table names to tables).
 */
export type DrizzleSchema = Record<string, DrizzleTable>;

/**
 * Drizzle introspector for converting Drizzle schemas to OrmAI metadata.
 */
export class DrizzleIntrospector {
  private readonly schema: DrizzleSchema;
  private readonly relations: Record<string, DrizzleRelation[]>;

  constructor(schema: DrizzleSchema, relations?: Record<string, DrizzleRelation[]>) {
    this.schema = schema;
    this.relations = relations ?? {};
  }

  /**
   * Convert Drizzle schema to OrmAI SchemaMetadata.
   */
  introspect(): SchemaMetadata {
    const models: Record<string, ModelMetadata> = {};

    for (const [tableName, table] of Object.entries(this.schema)) {
      if (!table._ || !table._.columns) {
        continue;
      }

      const modelName = this.tableNameToModelName(tableName);
      const fields: Record<string, FieldMetadata> = {};
      let primaryKey = 'id';
      const primaryKeys: string[] = [];

      // Process columns
      for (const [columnName, column] of Object.entries(table._.columns)) {
        fields[columnName] = {
          name: columnName,
          fieldType: this.mapColumnType(column.dataType, column.columnType),
          nullable: !column.notNull,
          primaryKey: column.primary ?? false,
          default: column.default,
        };

        if (column.primary) {
          primaryKey = columnName;
          primaryKeys.push(columnName);
        }
      }

      // Process relations
      const relations: Record<string, RelationMetadata> = {};
      const tableRelations = this.relations[tableName] ?? [];
      for (const rel of tableRelations) {
        relations[rel.fieldName] = {
          name: rel.fieldName,
          relationType: rel.relationType === 'many' ? 'one_to_many' : 'many_to_one',
          targetModel: this.tableNameToModelName(rel.referencedTableName),
          foreignKey: rel.fieldName,
        };
      }

      models[modelName] = {
        name: modelName,
        tableName: table._.name || tableName,
        fields,
        relations,
        primaryKey,
        ...(primaryKeys.length > 1 ? { primaryKeys } : {}),
      };
    }

    return { models };
  }

  /**
   * Convert table name to model name (PascalCase).
   */
  private tableNameToModelName(tableName: string): string {
    return tableName
      .split('_')
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
      .join('');
  }

  /**
   * Map Drizzle column type to OrmAI field type.
   */
  private mapColumnType(dataType: string, columnType?: string): FieldType {
    const dt = (columnType || dataType).toLowerCase();

    // String types
    if (dt.includes('varchar') || dt.includes('char') || dt.includes('text') || dt === 'string') {
      return 'string';
    }

    // Integer types
    if (
      dt.includes('int') ||
      dt === 'serial' ||
      dt === 'smallserial' ||
      dt === 'bigserial' ||
      dt === 'integer' ||
      dt === 'number'
    ) {
      return 'integer';
    }

    // Float types
    if (
      dt.includes('float') ||
      dt.includes('double') ||
      dt.includes('decimal') ||
      dt.includes('numeric') ||
      dt.includes('real')
    ) {
      return 'float';
    }

    // Boolean
    if (dt.includes('bool')) {
      return 'boolean';
    }

    // Date/Time
    if (dt.includes('timestamp') || dt.includes('datetime')) {
      return 'datetime';
    }
    if (dt.includes('date')) {
      return 'date';
    }
    if (dt.includes('time')) {
      return 'time';
    }

    // JSON
    if (dt.includes('json')) {
      return 'json';
    }

    // UUID
    if (dt.includes('uuid')) {
      return 'uuid';
    }

    // Binary - map to string since bytes is not a valid type
    if (dt.includes('bytea') || dt.includes('blob') || dt.includes('binary')) {
      return 'string';
    }

    // Default
    return 'string';
  }
}

/**
 * Create a DrizzleIntrospector from a Drizzle schema.
 */
export function createDrizzleIntrospector(
  schema: DrizzleSchema,
  relations?: Record<string, DrizzleRelation[]>
): DrizzleIntrospector {
  return new DrizzleIntrospector(schema, relations);
}
